src/k8s/pkg/component/dns.go

```diff
@@ -0,0 +1,107 @@
+package component
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	api "github.com/canonical/k8s/api/v1"
+	"github.com/canonical/k8s/pkg/utils"
+)
+
+func EnableDNSComponent(request api.UpdateDNSComponentRequest) error {
```

Comment:
A minor note on this. `EnableDNSComponent` is internal machinery, whereas `api.UpdateDNSComponentRequest` is "public" API.

If we had to change the API in any way (or this implementation), we don't want to have to make this change in multiple places. As such, I would recommend changing the definition to:

```
func EnableDNSComponent(upstreamNameserver, clusterDomain, serviceIP string) error
```

This way, we can change the API (or the EnableDNSComponent) without having to track down changes across multiple files.